// @magiclabs/ui/dist/cjs/components/layout/popover/popover.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _tslib = require('../../../_virtual/_tslib.js');
var composeRefs = require('@seznam/compose-react-refs');
var framerMotion = require('framer-motion');
var React = require('react');
var transitions = require('../../../hooks/transitions.js');
var useId = require('../../../hooks/use-id.js');
var useIsomorphicLayoutEffect = require('../../../hooks/use-isomorphic-layout-effect.js');
var clsx = require('clsx');
var exceptions = require('../../../libs/exceptions.js');
var forwardRefWrapper = require('../../../libs/forward-ref-wrapper.js');
var mergeProps = require('../../../libs/merge-props.js');
require('blueimp-md5');
require('color');
require('@artsy/fresnel');
var useTheme = require('../../theme-provider/use-theme.js');
var themeProvider = require('../../theme-provider/theme-provider.js');
require('../../theme-provider/theme-head.js');
var autoResizingDiv = require('../auto-resizing-div.js');
var ephemeralPortal = require('../ephemeral-portal.js');
var usePopperWrapper = require('./use-popper-wrapper.js');
var popover = require('./popover.less.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var composeRefs__default = /*#__PURE__*/_interopDefaultLegacy(composeRefs);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var clsx__default = /*#__PURE__*/_interopDefaultLegacy(clsx);

/* eslint-disable no-nested-ternary */
const PopoverBase = /* @__PURE__ */ forwardRefWrapper.forwardRefWrapper('Popover', {
    in: false,
    arrow: true,
    placement: 'bottom',
    strategy: 'absolute',
    portalize: false,
    appearance: 'default',
    stretch: false,
}, (props, externalRef) => {
    var _a, _b;
    const { children, in: inProp, portalize, placement, arrow, strategy, offset, appearance, stretch, className } = props, otherProps = _tslib.__rest(props, ["children", "in", "portalize", "placement", "arrow", "strategy", "offset", "appearance", "stretch", "className"]);
    const popper = usePopperWrapper.usePopperWrapper({ arrow, placement: placement, strategy: strategy, offset });
    const passthroughProps = {
        _portalize: portalize,
        _in: inProp,
        _hasArrow: arrow,
        _popper: popper,
        _appearance: appearance,
        _stretch: stretch,
    };
    const anchor = (_a = React__default["default"].Children.map(children, (child) => {
        if (React__default["default"].isValidElement(child) && child.type === PopoverAnchor) {
            return React__default["default"].cloneElement(child, mergeProps.mergeProps(child.props, Object.assign({}, passthroughProps)));
        }
        return null;
    })) === null || _a === void 0 ? void 0 : _a.filter(Boolean);
    const content = (_b = React__default["default"].Children.map(children, (child) => {
        if (React__default["default"].isValidElement(child) && child.type === PopoverContent) {
            return React__default["default"].cloneElement(child, mergeProps.mergeProps(child.props, Object.assign({}, passthroughProps)));
        }
        return null;
    })) === null || _b === void 0 ? void 0 : _b.filter(Boolean);
    // Validate that only one <Popover.Anchor> is received.
    const numAnchors = React__default["default"].Children.count(anchor);
    if (numAnchors > 1) {
        throw exceptions.createErrorWithCode('POPOVER_ONLY_ONE_ANCHOR_ALLOWED', PopoverBase, `Only one <Popover.Anchor> element can be rendered in a <Popover> at a time. Received: ${numAnchors}`);
    }
    // Validate that only one <Popover.Content> is received.
    const numContents = React__default["default"].Children.count(content);
    if (numContents > 1) {
        throw exceptions.createErrorWithCode('POPOVER_ONLY_ONE_CONTENT_ALLOWED', PopoverBase, `Only one <Popover.Content> element can be rendered in a <Popover> at a time. Received: ${numContents}`);
    }
    if (React__default["default"].Children.count(children) > numAnchors + numContents) {
        // Validate that only <Popover.Content> and <Popover.Anchor> components are received as children.
        throw exceptions.createErrorWithCode('POPOVER_INCOMPATIBLE_CHILDREN_DETECTED', PopoverBase, 'Only children wrapped with <Popover.Anchor> or <Popover.Content> are allowed within a <Popover>.');
    }
    return (React__default["default"].createElement("div", Object.assign({ className: clsx__default["default"]([popover["default"].Popover, className]) }, otherProps, { ref: externalRef }),
        anchor,
        content));
});
const PopoverAnchor = /* @__PURE__ */ forwardRefWrapper.forwardRefWrapper('Popover.Anchor', (props, externalRef) => {
    const _a = props, { children, _in, _portalize, _hasArrow, _popper, _appearance, _stretch } = _a, otherProps = _tslib.__rest(_a, ["children", "_in", "_portalize", "_hasArrow", "_popper", "_appearance", "_stretch"]);
    return (React__default["default"].createElement("div", Object.assign({}, otherProps, { ref: composeRefs__default["default"](_popper.anchorProps.ref, externalRef) }), children));
});
function getPopoverContentTransitionRotation(placement) {
    switch (placement) {
        case 'top':
            return { rotateX: '-7deg', rotateY: '0deg', transformOrigin: 'bottom center' };
        case 'top-start':
            return { rotateX: '7deg', rotateY: '7deg', transformOrigin: 'bottom left' };
        case 'top-end':
            return { rotateX: '7deg', rotateY: '-7deg', transformOrigin: 'bottom right' };
        case 'right':
            return { rotateX: '0deg', rotateY: '7deg', transformOrigin: 'left center' };
        case 'right-start':
            return { rotateX: '-7deg', rotateY: '7deg', transformOrigin: 'left top' };
        case 'right-end':
            return { rotateX: '7deg', rotateY: '7deg', transformOrigin: 'left bottom' };
        case 'left':
            return { rotateX: '0deg', rotateY: '-7deg', transformOrigin: 'right center' };
        case 'left-start':
            return { rotateX: '-7deg', rotateY: '-7deg', transformOrigin: 'right top' };
        case 'left-end':
            return { rotateX: '7deg', rotateY: '-7deg', transformOrigin: 'right bottom' };
        case 'bottom':
            return { rotateX: '-7deg', rotateY: '0deg', transformOrigin: 'top center' };
        case 'bottom-start':
            return { rotateX: '-7deg', rotateY: '7deg', transformOrigin: 'top left' };
        case 'bottom-end':
            return { rotateX: '-7deg', rotateY: '-7deg', transformOrigin: 'top right' };
        default:
            return { rotateX: '0deg', rotateY: '0deg', transformOrigin: 'center' };
    }
}
const popoverContentTransition = /* @__PURE__ */ transitions.createFramerTransition().withVariants({
    initial: ({ placement }) => (Object.assign(Object.assign({ opacity: 0, scale: 0.96 }, getPopoverContentTransitionRotation(placement)), { transition: { type: 'tween', ease: 'easeInOut', duration: 0.2 }, perspective: '500px' })),
    animate: () => ({
        opacity: 1,
        scale: 1,
        rotateX: '0deg',
        rotateY: '0deg',
        transition: { type: 'tween', ease: 'easeInOut', duration: 0.2 },
        perspective: '500px',
    }),
});
const PopoverContent = /* @__PURE__ */ forwardRefWrapper.forwardRefWrapper('Popover.Content', (props, externalRef) => {
    var _a, _b;
    const _c = props, { children, className, _in, _portalize, _hasArrow, _popper, _appearance, _stretch } = _c, otherProps = _tslib.__rest(_c, ["children", "className", "_in", "_portalize", "_hasArrow", "_popper", "_appearance", "_stretch"]);
    const id = useId.useID();
    const theme = useTheme.useTheme();
    const getPopoverContentTransitionProps = popoverContentTransition.use({
        initial: 'initial',
        animate: 'animate',
        exit: 'initial',
    });
    useIsomorphicLayoutEffect.useIsomorphicLayoutEffect(() => {
        var _a;
        (_a = _popper.forceUpdate) === null || _a === void 0 ? void 0 : _a.call(_popper);
    }, [_in]);
    const content = (React__default["default"].createElement("div", Object.assign({ className: clsx__default["default"]([
            popover["default"].PopoverContainer,
            _stretch && popover["default"][((_a = _popper.state) === null || _a === void 0 ? void 0 : _a.placement.match(/(top|bottom)/)) ? 'stretchHorizontal' : 'stretchVertical'],
            _hasArrow && _appearance === 'default' && popover["default"].hasArrow,
        ]) }, _popper.contentProps),
        React__default["default"].createElement(framerMotion.AnimatePresence, null, _in && (React__default["default"].createElement(framerMotion.motion.div, Object.assign({ key: id, className: clsx__default["default"]([popover["default"].PopoverContent, _appearance === 'none' && popover["default"].appearanceNone, className]) }, otherProps, getPopoverContentTransitionProps({ placement: (_b = _popper.state) === null || _b === void 0 ? void 0 : _b.placement }), { ref: externalRef }),
            React__default["default"].createElement(autoResizingDiv.AutoResizingDiv, { observerOnly: true, onResize: _popper.update }, children))))));
    return _portalize ? (React__default["default"].createElement(ephemeralPortal.EphemeralPortal, { className: popover["default"].PopoverPortal },
        React__default["default"].createElement(themeProvider.ThemeProvider, { theme: theme },
            React__default["default"].createElement("div", { className: popover["default"].Popover }, content)))) : (content);
});
const Popover = /* @__PURE__ */ Object.assign(PopoverBase, { Anchor: PopoverAnchor, Content: PopoverContent });

exports.Popover = Popover;
