// @magiclabs/ui/dist/es/libs/theme/default-theme.js
import { createTheme } from './create-theme.js';

const DEFAULT_THEME = /* @__PURE__ */ createTheme({
    primaryColor: {
        base: '#6851ff',
        lighter: '#a796ff',
        lightest: '#f1eeff',
        darker: '#4e3bdb',
        darkest: '#3728b7',
    },
    v2secondaryColor: {
        base: '#edebff',
        lighter: '#ffffff',
        lightest: '#ffffff',
        darker: '#e1dfff',
        darkest: '#ffffff',
    },
    v2tertiaryColor: {
        base: '#ffffff',
        lighter: '#ffffff',
        lightest: '#ffffff',
        darker: '#f8f8fa',
        darkest: '#ffffff',
    },
    neutralColor: {
        base: '#f8f8fa',
        lighter: '#ffffff',
        lightest: '#ffffff',
        darker: '#4e4d52',
        darkest: '#ffffff',
    },
    negativeColor: {
        base: '#d43101',
        lighter: '#ffffff',
        lightest: '#ffffff',
        darker: '#c92f00',
        darkest: '#ffffff',
    },
    lightColor: '#ffffff',
    darkColor: '#000000',
    midColor: {
        base: '#757575',
        lighter: '#bdbdbd',
        lightest: '#eeeeee',
        darker: '#545454',
        darkest: '#212121',
    },
    successColor: {
        base: '#00c988',
        lighter: '#5feea4',
        lightest: '#ddfde6',
        darker: '#00ac85',
        darkest: '#00907e',
    },
    warningColor: {
        base: '#bf9046',
        lighter: '#ebd092',
        lightest: '#fbf5df',
        darker: '#a47333',
        darkest: '#895923',
    },
    errorColor: {
        base: '#ff5635',
        lighter: '#FBA67F',
        lightest: '#FAE6DC',
        darker: '#d02f20',
        darkest: '#B71C1A',
    },
    typography: {
        h1: {
            fontSize: '2.5rem',
            lineHeight: '3.5rem',
            defaultFontWeight: '800',
        },
        h2: {
            fontSize: '2rem',
            lineHeight: '2.75rem',
            defaultFontWeight: '800',
        },
        h3: {
            fontSize: '1.5rem',
            lineHeight: '2.25rem',
            defaultFontWeight: '700',
        },
        h4: {
            fontSize: '1.25rem',
            lineHeight: '2rem',
            defaultFontWeight: '600',
        },
        h5: {
            fontSize: '1.25rem',
            lineHeight: '2rem',
            defaultFontWeight: '400',
        },
        h6: {
            fontSize: '.75rem',
            lineHeight: '.75rem',
            defaultFontWeight: '600',
            textTransform: 'uppercase',
        },
        p: {
            large: {
                fontSize: '1.125rem',
                lineHeight: '1.75rem',
                defaultFontWeight: '600',
            },
            medium: {
                fontSize: '1rem',
                lineHeight: '1.5rem',
                defaultFontWeight: '600',
            },
            small: {
                fontSize: '.875rem',
                lineHeight: '1.5rem',
                defaultFontWeight: '600',
            },
        },
        mono: {
            large: {
                fontSize: '1.125rem',
                lineHeight: '1.75rem',
                fontFamily: 'var(--mg__font_monospace)',
            },
            medium: {
                fontSize: '1rem',
                lineHeight: '1.5rem',
                fontFamily: 'var(--mg__font_monospace)',
            },
            small: {
                fontSize: '.875rem',
                lineHeight: '1.5rem',
                fontFamily: 'var(--mg__font_monospace)',
            },
        },
    },
});
function overrideDefaultTheme(config) {
    return merge(config, merge(DEFAULT_THEME.config, {}));
}
function merge(source, target) {
    for (const [key, val] of Object.entries(source)) {
        if (val !== null && typeof val === `object`) {
            // @ts-ignore
            if (target[key] === undefined) {
                // @ts-ignore
                target[key] = {};
            }
            // @ts-ignore
            merge(val, target[key]);
        }
        else {
            // @ts-ignore
            target[key] = val;
        }
    }
    return target;
}

export { DEFAULT_THEME, overrideDefaultTheme };
