// @magiclabs/ui/dist/cjs/components/layout/inset/inset.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var useStyles = require('../../../hooks/use-styles.js');
var clsx = require('clsx');
var mergeProps = require('../../../libs/merge-props.js');
var inset = require('./inset.less.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var clsx__default = /*#__PURE__*/_interopDefaultLegacy(clsx);

function resolveStyleValue(source) {
    return typeof source === 'number' ? `${source}px` : source;
}
const Inset = (props) => {
    const { wrapItemsWith, top, right, bottom, left, all, trim, children } = props;
    const style = useStyles.useStyles(all
        ? {
            padding: resolveStyleValue(all),
        }
        : {
            paddingTop: resolveStyleValue(top),
            paddingRight: resolveStyleValue(right),
            paddingBottom: resolveStyleValue(bottom),
            paddingLeft: resolveStyleValue(left),
        });
    const numChildren = React__default["default"].Children.count(children);
    const wrappedChildren = React__default["default"].Children.map(children, (child, i) => {
        const isFirstChild = i === 0;
        const isLastChild = i === numChildren - 1;
        const className = clsx__default["default"]([
            inset["default"].Inset,
            isFirstChild && inset["default"].isFirstChild,
            isLastChild && inset["default"].isLastChild,
            trim === 'horizontal' && inset["default"].trimHorizontally,
            trim === 'vertical' && inset["default"].trimVertically,
        ]);
        if (wrapItemsWith) {
            return React__default["default"].createElement(wrapItemsWith, { className, style }, child);
        }
        if (React__default["default"].isValidElement(child)) {
            return React__default["default"].cloneElement(child, mergeProps.mergeProps(child.props, { className, style }));
        }
        return child;
    });
    return React__default["default"].createElement(React__default["default"].Fragment, null, wrappedChildren);
};

exports.Inset = Inset;
